File: Minimax.java
/**
 * Class that contains the MINIMAX Algorithm
 */
public class Minimax implements Constants{

    private Node initialNode;
    public int nTotalNodes = 0;

    Minimax(){}

    /**
     * Minimax algorithm
     *
     * @param node current node of the current game board
     * @param depth chosen by the user at the start
     * @param player current player
     * @return the best column for player to play
     */
    public int min_max(Node node, int depth, int player){
        this.initialNode = node;

        int value = Integer.MAX_VALUE;
        int bestMove = 0;

        for(Node child : node.getSuccessors(PC)){
            nTotalNodes++;
            int v = minFunction_MinMax(child, depth - 1, player);
            if(v <= value){
                value = v;
                bestMove = child.getColumn();
            }
        }

        this.initialNode.setNumTotalNodes(nTotalNodes);

        return bestMove;
    }

    /**
     * Auxiliar function of the minimax algorithm, that minimizes the player
     *
     * @param node current node of the current game board
     * @param depth chosen by the user at the start
     * @param player current player
     * @return the node's best utility value
     */
    private int minFunction_MinMax(Node node, int depth, int player){
        if(depth == 0 || node.getState().isGameOver(player) || node.getState().checkDraw())
            return node.getState().utility(player);

        int v  = Integer.MAX_VALUE;

        for(Node child : node.getSuccessors(PLAYER1)) {
            nTotalNodes++;
            v = Math.min(v, maxFunction_MinMax(child, depth - 1, player));
        }

        return v;
    }

    /**
     * Auxiliar function of the minimax algorithm, that maximizes the player
     *
     * @param node current node of the current game board
     * @param depth chosen by the user at the start
     * @param player current player
     * @return the node's best utility value
     */
    private int maxFunction_MinMax(Node node, int depth, int player){
        if(depth == 0 || node.getState().isGameOver(player) || node.getState().checkDraw())
            return node.getState().utility(player);

        int v  = Integer.MIN_VALUE;

        for(Node child : node.getSuccessors(PC)) {
            nTotalNodes++;
            v = Math.max(v, minFunction_MinMax(child, depth - 1, player));
        }

        return v;
    }

}

File: js/Minimax.js
var initialNode;


function minimax(node, depth, player){
    initialNode = node;

    var value = Number.MAX_SAFE_INTEGER;
    var bestMove = 0;

    let children = node.getSuccessors("PC");
    for(var child of children){
        var v = maxFunction_MiniMax(child, depth - 1, "PC");
        if(v <= value){
            value = v;
            bestMove = child.getColumn();
        }
    }

    //this.initialNode.setNumTotalNodes(nTotalNodes);
    return bestMove;
}

function minFunction_MiniMax(node, depth, player){
    if(depth == 0 || isGameOver(node.getState(), player) || checkDraw(node.getState()))
        return utility(node.getState(), player);

    var v  = Number.MAX_SAFE_INTEGER;

    let children = node.getSuccessors("PC");
    for(var child of children) {
        v = Math.min(v, maxFunction_MiniMax(child, depth - 1, "PC"));
    }
    return v;
}

function maxFunction_MiniMax(node, depth, player){
    if(depth == 0 || isGameOver(node.getState(), player) || checkDraw(node.getState()))
        return utility(node.getState(), player);

    var v  = Number.MIN_SAFE_INTEGER;

    let children = node.getSuccessors("PLAYER");
    for(var child of children) {
        v = Math.max(v, minFunction_MiniMax(child, depth - 1, "PLAYER"));
    }
    return v;
}